python/ray/workflow/workflow_storage.py

```diff
@@ -167,6 +165,10 @@ def save_step_output(self, step_id: StepID, ret: Union[Workflow, Any], *,
                 #     self._put(self._key_step_exception(step_id), exception))
 
         asyncio_run(asyncio.gather(*tasks))
+        if dynamic_output_id:
+            asyncio_run(
+                self._update_dynamic_output(outer_most_step_id,
+                                            dynamic_output_id))
```

Comment:
Should we put it to tasks and just call `        asyncio_run(asyncio.gather(*tasks))`?

Reply:
There is a race condition: if we update the dynamic output before finishing checkpointing, then during recovery, the dynamic output would point to a checkpoint that does not exist. Let me comment in the code.